### ISSUE TITLE
No apparent way to check for a valid service endpoint; incorrect documentation on Amazon.RegionEndpoint.GetEndpointForService

### ISSUE THREAD

**Issue by philipflesher** [aws/aws-sdk-net#868]
The documentation for `Amazon.RegionEndpoint.GetEndpointForService(string)` indicates that it throws `ArgumentException` if the passed service name has no valid endpoint for the region. Via both testing and inspection of the code, this is clearly not the case. Instead, a fabricated endpoint is constructed and returned, which may or may not be valid. This documentation should be corrected.

I see no other way in the codebase to check whether a request will succeed to a given region endpoint without making an attempted request, which necessitates DNS resolution and a potentially very long wait. Is there a way to check for a valid service endpoint for a given region?

Assuming there is no way to do so currently, could a new method be implemented on `RegionEndpoint`, similar to `GetEndpointForService`, that performs this necessary operation?

**Comment by philipflesher** [comment]
The below code snippet (C#) should show the issue.

```
            // Helper method to get credentials; use any credentials source
            var creds = Aws.AwsCredentials.FromSharedCredentialsFile("peopleadmin-sfngprod");

            // Configure a client against a region/service combo that is not (yet?) valid.
            // Example: WorkSpaces is not currently available in ap-south-1
            var apSouth1 = Amazon.RegionEndpoint.APSouth1;
            var workspacesClient = new Amazon.WorkSpaces.AmazonWorkSpacesClient(creds, apSouth1);

            // We could check it first, if the following method did what its documentation claims...
            try
            {
                apSouth1.GetEndpointForService(workspacesClient.Config.RegionEndpointServiceName);
            }
            catch (ArgumentException)
            {
                // Invalid region
                throw new Exception("Caught the exception we'd hope to get.");
            }

            // Throws an exception after a long HTTP delay attempting to resolve the hostname
            workspacesClient.DescribeWorkspacesAsync().Wait();
```

**Comment by sstevenkang** [comment]
Sorry about the confusion.  We will definitely correct the documentation.  It's an interesting idea you are proposing, but I am not sure how you would guarantee that region endpoint is really available without making a HTTP request.

SDK currently relies on [endpoints.json](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/endpoints.json) file.  This gets periodically updated with a release. The only guarantee we make is that whatever is in this file **is** reachable.  So, if we were to implement a utility function that checks using this file, you'll get false negatives.  On the other hand, if we disable the function of generating a possible endpoint, customers who want to reach new regions will be forced to update the nuget package to get the latest endpoints.json--which may contain breaking changes.

This behavior is not unique to .NET SDK.  All AWS SDK uses heuristics to calculate endpoints to maintain forward compatibility.  The authoritative source of which regions are available is [here](https://docs.aws.amazon.com/general/latest/gr/rande.html).

**Comment by philipflesher** [comment]
Thanks for the quick response @sstevenkang!

Yes, that forward compatibility makes perfect sense as the motivation for the current behavior, and I divined as much in the past. In fact, this is the first time I've had a use case where a helper method to do what I'm trying to do would be useful - in the past, forward compatibility was probably more desired for me.

With appropriate documentation, and a proper method name (e.g., (`CheckValidServiceNonForwardCompatible`), would it be possible to implement a non-forward-compatible method that does what I'm suggesting? This would be extremely useful in the current situation I'm in, where we regularly need to make checks against all resources, in all regions, in all services - and these HTTP delays are a killer. The false negatives are not much of an issue for us with this use case.

We could code this up separately, and try to keep our own record of valid endpoints current, but this feels like a waste of effort when the SDK already keeps a fairly current record itself and we can update our package as we please.

**Comment by philipflesher** [comment]
Alternatively - do you have any suggestion for how I might hack my way into an implementation that relies on the internal JSON? There is some interesting stuff going on in RegionEndpointProviderV3, but I don't see a way to utilize it to achieve my ends.

**Comment by sstevenkang** [comment]
It's a difficult request to fulfill because `endpoints.json` is not a well defined format.  There's no guarantee that the format will not change and what information maybe expoed/removed. If we provide that utility function, AWS needs to somehow maintain backward compatibility and support that functionality until the next version of the SDK.  We explicitly made the decision to make a lot of the functions on `RegionEndpointProviderV3` private for this reason.

If HTTP delay is a bottleneck, would parallel-izing that work help?  It's a bit obfuscated, but `RegionEndpointProviderV3` should contain most of the logic on how to read and parse the information in `endpoints.json` file.

**Comment by philipflesher** [comment]
How does the Java SDK manage to provide this functionality? Can we do the same?

https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-region-selection.html

**Comment by philipflesher** [comment]
In answer to your other question, parallelizing something with a 20- to 30-second delay isn't an option for us. The only way I can see us getting around this at the moment is to code up a corollary to the RegionEndpointProvider also keeps a global cache of reachable endpoints, which it populates after attempting API calls. And maybe we have to route all API requests through it... yuck.

**Comment by sstevenkang** [comment]
Java SDK has a very similar logic.

**Comment by philipflesher** [comment]
Right, similar, except that it actually exposes what I'm asking for here:

https://github.com/aws/aws-sdk-java/blob/4734de6fb0f80fe5768a6587aad3b9d0eaec388f/aws-java-sdk-core/src/main/java/com/amazonaws/regions/InMemoryRegionImpl.java#L83

Or am I misreading that?

**Comment by philipflesher** [comment]
@normj this is definitely still a relevant issue for us. How do I ensure it gets more activity? :)

Also, it was previously mentioned that the method docs would be updated, and I haven't checked back on that, but I see no PR merge referencing this issue.

**Comment by normj** [comment]
Sorry I though the doc changed has happened. I just made the doc change. https://github.com/aws/aws-sdk-net/commit/5830d2e839672802b611882d656652418d7de819

I really don't want to expose any more of the regional functionality then we already are. To deal with the huge growth in AWS Services we have changed the endpoint logic many times over the years and I suspect we will change again in the future. If I expose anymore functionality that makes it hard for us to make changes with out causing breaking changes.

If you really want this type of information you can always pull out the endpoint.json file as an embedded resource from AWSSDK.Core and run your own logic. Keep in mind though we aren't making any guarantees that we don't change to different system in the future.

**Comment by normj** [comment]
I agree a service catalog would be great. We have an internal team that owns that endpoints.json file and we have encouraged them to turn it into a service but they are not ready for that. I'm not sure if and when that will happen.

To be honest I'm not sure where you can add your voice as it is one of those global meta feature request that doesn't have a home. But we do bring these request up with the team that owns the region and endpoint metadata.
